core/src/main/java/org/elasticsearch/cluster/metadata/IndexGraveyard.java

```diff
@@ -0,0 +1,466 @@
+import java.util.function.BiFunction;
+
+/**
+ * A collection of tombstones for explicitly marking indices as deleted in the cluster state.
+ *
+ * The cluster state contains a list of index tombstones for indices that have been
+ * deleted in the cluster.  Because cluster states are processed asynchronously by
+ * nodes and a node could be removed from the cluster for a period of time, the
+ * tombstones remain in the cluster state for a fixed period of time, after which
+ * they are purged.
+ */
+final public class IndexGraveyard implements MetaData.Custom {
+
+    /**
+     * Setting for the maximum tombstones allowed in the cluster state;
+     * prevents the cluster state size from exploding too large, but it opens the
+     * very unlikely risk that if there are greater than MAX_TOMBSTONES index
+     * deletions while a node was offline, when it comes back online, it will have
+     * missed index deletions that it may need to process.
+     */
+    public static final Setting<Integer> SETTING_MAX_TOMBSTONES = Setting.intSetting("cluster.indices.tombstones.size",
+                                                                                     500, // the default maximum number of tombstones
+                                                                                     Setting.Property.NodeScope);
+
+    public static final IndexGraveyard PROTO = new IndexGraveyard(new ArrayList<>());
+    public static final String TYPE = "index-graveyard";
+    private static final ParseField TOMBSTONES_FIELD = new ParseField("tombstones");
+    private static final ObjectParser<List<Tombstone>, Void> GRAVEYARD_PARSER = new ObjectParser<>("index_graveyard", ArrayList::new);
+    static {
+        GRAVEYARD_PARSER.declareObjectArray(List::addAll, Tombstone.getParser(), TOMBSTONES_FIELD);
+    }
+
+    private final List<Tombstone> tombstones;
+
+    private IndexGraveyard(final List<Tombstone> list) {
+        tombstones = Collections.unmodifiableList(list);
+    }
+
+    private IndexGraveyard(final StreamInput in) throws IOException {
+        final int queueSize = in.readVInt();
+        List<Tombstone> tombstones = new ArrayList<>(queueSize);
+        for (int i = 0; i < queueSize; i++) {
+            tombstones.add(new Tombstone(in));
+        }
+        this.tombstones = Collections.unmodifiableList(tombstones);
+    }
+
+    public static IndexGraveyard fromStream(final StreamInput in) throws IOException {
+        return new IndexGraveyard(in);
+    }
+
+    @Override
+    public String type() {
+        return TYPE;
+    }
+
+    @Override
+    public EnumSet<MetaData.XContentContext> context() {
+        return MetaData.API_AND_GATEWAY;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        return (obj instanceof IndexGraveyard) && Objects.equals(tombstones, ((IndexGraveyard)obj).tombstones);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(tombstones);
```

Comment:
Agreed, the class had more members which is why I used the `Objects#hash`, but now that it only has one member, it makes sense.